File: src/main/java/com/github/nnnnusui/format/Base32hex.java
package com.github.nnnnusui.format;

import java.nio.ByteBuffer;
import java.util.BitSet;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Base32hex {
    public static String encode(byte[] source) { return instance._encode(source); }
    public static byte[] decode(String source) { return instance._decode(source); }
    //    public static Base32hex getInstance() { return instance; }
    private static Base32hex instance = new Base32hex();
    private Base32hex() {}

    private final int radix = 32;
    private final int blockSize = 40; // bit per base32 block
    private final int wordSize  =  5; // bit per base32 word
    private final char fillCharacter = '=';
    private final int bitPerByte  = 8;

    public String _encode(byte[] _source){
        final int sourceBitQuantity = _source.length * bitPerByte;
        final int resultWordQuantity = ((sourceBitQuantity -1) / wordSize) +1;
        final int resultBitQuantity = ((sourceBitQuantity -1) / blockSize +1) * blockSize;
        final BitSet source = reverse(BitSet.valueOf(_source), bitPerByte);

        return IntStream.range(0, resultWordQuantity)
                .mapToObj(index-> source.get(index * wordSize, (index +1) * wordSize))
                .map(this::getWord)
                .map(it-> Integer.toString(it, radix))
                .map(String::valueOf)
                .collect(Collectors.joining())
                .toUpperCase()
            + Stream.generate(() -> Character.toString(fillCharacter))
                .limit((resultBitQuantity - sourceBitQuantity) / wordSize)
                .collect(Collectors.joining());
    }
    public byte[] _decode(String _source){
        final String source = _source.replace(Character.toString(fillCharacter), "");
        final ByteBuffer sourceBytes = ByteBuffer.allocate(source.length() * wordSize);
        source.chars()
                .mapToObj(it -> (char) it)
                .map(it -> Character.toString(it))
                .map(it -> Byte.parseByte(it, radix))
                .forEach(sourceBytes::put);

        final BitSet result = setWords(sourceBytes.array());
        return result.toByteArray();
    }

    private int getWord(BitSet bitSet){
        return IntStream.range(0, wordSize)
                .mapToObj(bitSet::get)
                .map(it-> it ? 1 : 0)
                .reduce(0, (sum, bit) -> (sum << 1) + bit); // Binary to decimal.
    }
    private BitSet setWords(byte[] values){ // TODO: Refactoring
        final BitSet bitSet = new BitSet(values.length * wordSize);
        IntStream.range(0, values.length)
                .forEach(index->
                    reverse(BitSet.valueOf(new byte[]{ values[index] }), wordSize)
                            .stream()
                            .map(it-> it + (index * wordSize))
                            .forEach(bitSet::set)
                );
        return reverse(bitSet, bitPerByte);
    }

    private BitSet reverse(BitSet source, int unitSize){
        final BitSet result = new BitSet(source.length());
        source.stream()
                .forEach(it-> {
                    int multiplesOfUnit = (it / unitSize) * unitSize;
                    int lowerThanUnit = (unitSize -1) - (it % unitSize);
                    result.set(multiplesOfUnit + lowerThanUnit);
                });
        return result;
    }
}

File: src/main/scala/com/github/nnnnusui/format/Base32hex.scala
package com.github.nnnnusui.format

import java.nio.ByteBuffer
import java.util

object Base32hex {
  def encode(source: Array[Byte]): String      = _encode(source)
  def decode(source: String     ): Array[Byte] = _decode(source)

  private val bitPerByte =  8
  private val radix      = 32
  private val blockSize  = 40 // bit per base32 block
  private val wordSize   =  5 // bit per base32 word
  private val fillCharacter = '='


  private def _encode(_source: Array[Byte]): String ={
    val sourceBitQuantity  = _source.length * bitPerByte
    val resultWordQuantity = ceil(sourceBitQuantity, from = wordSize)
    val resultBitQuantity  = resultWordQuantity * wordSize
    val fillCharacterQuantity = ceil(resultBitQuantity - sourceBitQuantity, from = wordSize)
    val source = util.BitSet.valueOf(_source).reverse(bitPerByte)

    Range(0, resultWordQuantity)
        .map(index=> source.get(index * wordSize, (index +1) * wordSize))
        .map(getWord)
        .map(it=> Integer.toString(it, radix))
        .mkString
        .toUpperCase() +
      fillCharacter.toString * fillCharacterQuantity
  }
  private def _decode(_source: String): Array[Byte] ={
    val source = _source.replace(fillCharacter.toString, "")
    val sourceBytes = ByteBuffer.allocate(source.length * wordSize)
    source.toCharArray
      .map(_.toString)
      .map(it=> BigInt(it, radix))
      .map(_.toByte)
      .foreach(sourceBytes.put)
    setWords(sourceBytes.array())
      .toByteArray
  }

  def ceil(x: Int, from: Int): Int =
    (x / from) + (if (x % from == 0) 0 else 1)
  private def getWord(bitSet: util.BitSet): Int =
    Range(0, wordSize)
    .map(bitSet.get)
    .map(it=> if(it) 1 else 0)
    .fold(0)((sum, bit)=> (sum << 1) + bit)
  private def setWords(values: Array[Byte]): util.BitSet ={
    val bitSet = new util.BitSet(values.length * wordSize)
    values.indices
      .foreach{index=>
        util.BitSet.valueOf(Array(values(index))).reverse(wordSize)
          .stream()
          .map(_ + (index * wordSize))
          .forEach(it=> bitSet.set(it))
      }
    bitSet.reverse(bitPerByte)
  }
  implicit class RichBitSet(val src: util.BitSet){
    def reverse(unitSize: Int): util.BitSet = {
      val result = new util.BitSet(src.length)
      src.stream.forEach{ it=>
        val multiplesOfUnit = (it / unitSize) * unitSize
        val lowerThanUnit   = (unitSize - 1) - (it % unitSize)
        result.set(multiplesOfUnit + lowerThanUnit)
      }
      result
    }
  }
}
